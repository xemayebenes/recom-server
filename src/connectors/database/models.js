import Mongoose from 'mongoose';

import {
  UserSchema,
  options,
  ItemSchema,
  LastItemSchema,
  NotificationSchema
} from './schemas';

export const Notification = Mongoose.model('Notification', NotificationSchema);

export const Item = Mongoose.model('Item', ItemSchema);

export const LastItem = Mongoose.model('LastItem', LastItemSchema);

export const User = Mongoose.model('users', UserSchema);

export const Movie = Item.discriminator(
  'Movie',
  new Mongoose.Schema({}, options)
);
export const Serie = Item.discriminator(
  'Serie',
  new Mongoose.Schema({}, options)
);
