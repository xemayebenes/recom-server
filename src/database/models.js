import Mongoose from 'mongoose';

import { MOVIE_ITEM, SERIE_ITEM } from '../constants';

import {
  UserSchema,
  options,
  ItemSchema,
  LastItemSchema,
  NotificationSchema,
  ListSchema
} from './schemas';

export const Notification = Mongoose.model('Notification', NotificationSchema);

export const Item = Mongoose.model('Item', ItemSchema);

export const LastItem = Mongoose.model('LastItem', LastItemSchema);

export const User = Mongoose.model('users', UserSchema);

export const Movie = Item.discriminator(
  MOVIE_ITEM,
  new Mongoose.Schema({}, options)
);
export const Serie = Item.discriminator(
  SERIE_ITEM,
  new Mongoose.Schema({}, options)
);

export const List = Mongoose.model('List', ListSchema);
