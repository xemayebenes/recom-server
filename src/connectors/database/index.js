import Mongoose from 'mongoose';

const MOVIE_ITEM = 'Movie';
const SERIE_ITEM = 'Serie';
const Schema = Mongoose.Schema;

// import { Movie, Serie } from './models';

export default ({
    dbAddress = 'mongodb://localhost/recom',
}) => {

    const dataBase = {};


    const UserSchema = Mongoose.Schema({
        email: String,
        movies: [{ type: Schema.Types.ObjectId, ref: 'Item' }],
        series: [{ type: Schema.Types.ObjectId, ref: 'Item' }],

    });

    const options = { discriminatorKey: 'kind' };

    const ItemSchema = Mongoose.Schema({
        externalId: Number,
        user: { type: Schema.Types.ObjectId, ref: 'User' },
    }, options);

    const Item = Mongoose.model('Item', ItemSchema);

    const LastItemSchema = Mongoose.Schema({
        date: Date,
        kind: String,
        item: { type: Schema.Types.ObjectId, ref: 'Item' },
        user: { type: Schema.Types.ObjectId, ref: 'User' },
    });

    const LastItem = Mongoose.model('LastItem', LastItemSchema);

    const User = Mongoose.model('users', UserSchema);
    const Movie = Item.discriminator('Movie', new Mongoose.Schema({}, options));
    const Serie = Item.discriminator('Serie', new Mongoose.Schema({}, options));


    dataBase.connect = () => Mongoose.connect('mongodb://localhost/recom', {});

    dataBase.createMovie = async(externalId, userId) => {
        const movie = new Movie({ externalId, user: userId })
        await movie.save()
            .then(async() => {
                await User.findByIdAndUpdate(userId, { $push: { movies: movie } }, { safe: true, upsert: true });
                const lastItem = new LastItem({
                    date: new Date(),
                    kind: MOVIE_ITEM,
                    item: movie,
                    user: userId,
                });
                await lastItem.save();
                return;
            });
        return movie;
    };
    dataBase.createSerie = async(externalId, userId) => {
        const serie = new Serie({ externalId, user: userId })
        await serie.save()
            .then(async() => {
                await User.findByIdAndUpdate(userId, { $push: { series: serie } }, { safe: true, upsert: true });
                const lastItem = new LastItem({
                    date: new Date(),
                    kind: SERIE_ITEM,
                    item: serie,
                    user: userId,
                });
                await lastItem.save();
                return;
            });
        return serie;
    };
    const getItemByUser = (Model, userId) => Model.find({ user: userId });

    dataBase.getSeriesByUser = (userId) => getItemByUser(Serie, userId);

    dataBase.getMoviesByUser = (userId) => getItemByUser(Movie, userId);

    dataBase.getLastItemsByUser = (userId) => LastItem.find({ user: userId }).populate('item').sort({ date: 1 }).limit(10);

    return dataBase
};