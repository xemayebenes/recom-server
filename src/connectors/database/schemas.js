import Mongoose from 'mongoose';

const Schema = Mongoose.Schema;

export const UserSchema = Mongoose.Schema({
  email: String,
  password: String,
  movies: [{ type: Schema.Types.ObjectId, ref: 'Item' }],
  series: [{ type: Schema.Types.ObjectId, ref: 'Item' }],
  notifications: [{ type: Schema.Types.ObjectId, ref: 'Notification' }]
});

export const options = { discriminatorKey: 'type' };

export const ItemSchema = Mongoose.Schema(
  {
    externalId: Number,
    completed: Boolean,
    user: { type: Schema.Types.ObjectId, ref: 'User' },
    list: { type: Schema.Types.ObjectId, ref: 'List' }
  },
  options
);

export const LastItemSchema = Mongoose.Schema({
  date: Date,
  type: String,
  completed: Boolean,
  item: { type: Schema.Types.ObjectId, ref: 'Item' },
  user: { type: Schema.Types.ObjectId, ref: 'User' }
});

export const ListSchema = Mongoose.Schema({
  name: String,
  date: Date,
  items: [{ type: Schema.Types.ObjectId, ref: 'Item' }],
  user: { type: Schema.Types.ObjectId, ref: 'users' }
});

export const NotificationSchema = Mongoose.Schema({
  new: Boolean,
  user: { type: Schema.Types.ObjectId, ref: 'Users' },
  from: { type: Schema.Types.ObjectId, ref: 'users' },
  date: Date,
  type: String,
  title: String,
  externalId: Number
});
