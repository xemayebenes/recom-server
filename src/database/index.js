import Mongoose from 'mongoose';

import { User } from './models';

import listService from './services/lists.service';
import itemsService from './services/items.service';
import notificationsService from './services/notifications.service';

export default ({
  dbAddress = 'mongodb://recom:recom123@ds139251.mlab.com:39251/recom'
}) => {
  const dataBase = {};

  dataBase.connect = () =>
    Mongoose.connect(
      'mongodb://recom:recom123@ds139251.mlab.com:39251/recom',
      {}
    );

  dataBase.getUserByEmailPassword = (email, password) =>
    User.findOne({ email, password });

  dataBase.createUser = (email, password, user) => {
    const newUser = new User({ email, password, user });
    return newUser.save();
  };

  dataBase.getUserByEmail = email => User.findOne({ email });
  dataBase.getUserByUser = user => User.findOne({ user });

  dataBase.listService = listService;
  dataBase.itemsService = itemsService;
  dataBase.notificationsService = notificationsService;

  return dataBase;
};
