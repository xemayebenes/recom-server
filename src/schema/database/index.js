export const UserMovie = `
  type UserMovie {
    id: ID!
    externalId: Int
    film: Movie
  }
`;

export const UserSerie = `
type UserSerie {
  id: ID!
  externalId: Int
  serie: Serie
}
`;
export const LastItem = `
type LastItem {
  id: ID!
  date: Date
  type: String
  item: ItemInterface 
}
`;

export const Date = `
  scalar Date
`;