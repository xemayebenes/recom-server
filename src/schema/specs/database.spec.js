import { UserMovie, UserSerie, LastItem, Date } from '../database';

describe('SCHEMA DATA BASE', () => {
    describe('UserMovie', () => {
        it('should match snapshot', () => {
            expect(UserMovie).toMatchSnapshot();
        });
    });
    describe('UserSerie', () => {
        it('should match snapshot', () => {
            expect(UserSerie).toMatchSnapshot();
        });
    });
    describe('LastItem', () => {
        it('should match snapshot', () => {
            expect(LastItem).toMatchSnapshot();
        });
    });
    describe('Date', () => {
        it('should match snapshot', () => {
            expect(Date).toMatchSnapshot();
        });
    });
});