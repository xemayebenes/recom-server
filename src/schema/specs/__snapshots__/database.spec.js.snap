// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCHEMA DATA BASE Date should match snapshot 1`] = `
"
  scalar Date
"
`;

exports[`SCHEMA DATA BASE LastItem should match snapshot 1`] = `
"
type LastItem {
  id: ID!
  date: Date
  type: String
  item: ItemInterface 
}
"
`;

exports[`SCHEMA DATA BASE UserMovie should match snapshot 1`] = `
"
  type UserMovie {
    id: ID!
    externalId: Int
    film: Movie
  }
"
`;

exports[`SCHEMA DATA BASE UserSerie should match snapshot 1`] = `
"
type UserSerie {
  id: ID!
  externalId: Int
  serie: Serie
}
"
`;
